libnavigation-core/src/main/java/com/mapbox/navigation/core/arrival/ArrivalProgressObserver.kt

```diff
@@ -51,47 +52,62 @@ internal class ArrivalProgressObserver(
             ?: return
 
         val arrivalOptions = arrivalController.arrivalOptions()
-        val hasMoreLegs = hasMoreLegs(routeProgress)
-        if (routeProgress.currentState == RouteProgressState.ROUTE_COMPLETE && !hasMoreLegs) {
+        val arrivalProgress = toArrivalProgress(arrivalOptions, routeProgress, routeLegProgress)
```

Comment:
Are we ok creating an `ArrivalProgress` on every route progress update (every ~1s)? https://github.com/mapbox/mapbox-navigation-android/pull/4246#issuecomment-816966428

> Curious, is there a performance impact?
>
> If you want to test it, could try pulling this branch and making a ui-benchmark https://github.com/mapbox/mapbox-navigation-android/pull/4239

:trollface:

Reply:
lol the dream of the benchmarks! but my comment may have come across wrong, i didn't dig deep into the original optimization https://github.com/mapbox/mapbox-navigation-android/pull/4244 😄 

optimizing for being readable here, is there a memory perf concern?

Reply:
> is there a memory perf concern?

That's what I was wondering myself. Is there? Were we accessing directly to this data instead of bundling it into an object before? How much does that creating this object on every update tax in terms of memory perf?

Reply:
I did a little reading to refresh my memory. My brain doesn't care about memory like it used to.

in the old Android days, creating objects was a big concern. Because the garbage collector originally ran on a single thread (Dalvik), it would randomly freeze your phone to clean up objects like this. Here's a good overview https://proandroiddev.com/collecting-the-garbage-a-brief-history-of-gc-over-android-versions-f7f5583e433c

Our minimum SDK version is 21, which does not use Dalvik, it's using ART (Android Runtime). Big benefit is how the garbage collector works, it no longer needs to freeze the world to clean up memory. It also cleans up memory based on the time to live, newer phones have even more improvements.

Memory concerns arise if; there is a leak, the time to live is forever, or very large allocations. This object dies right after it is created, so it won't be an issue.

Unrelated but, have you read about Rust? It solves memory issues at the [compiler level](https://doc.rust-lang.org/book/ch04-00-understanding-ownership.html).